README.md

## Parallel SSH with Regex

Usage:

    pypsh hostregex cmd

E.g.:

    pypsh "role\d+\.customer\.your\.domain" "uptime"

This matches every host in the `known_hosts` file against the regex and executes the command.


## Installation:

    pip install pypsh 
    
## Development:

    git clone https://github.com/mfussenegger/pypsh.git
    cd pypsh
    mkvirtualenv pypsh
    pip install -r requirements.txt


CHANGES.rst
=================
Changes for pypsh
=================

Unreleased
==========

2015-07-13: 0.7.3
=================

 - detect hosts from Host entries in `~/.ssh/config`

 - fixed support for ``--pty``

2014-05-01: 0.7.2
=================

 - fixed matching of hosts that have a custom port configured.

 - added the `--pty` option.

 - stdout or stderr buffers filling up should no longer cause pypsh to get stuck

2014-03-14: 0.7.1
=================

 - **bugfix**: adopt paramiko missing_host_key usage to changed API

2014-03-14: 0.7.0
=================

 - **breaking change**
   Changed the commandline interface again. The `hostregex` is now always the
   first (required) parameter. In addition the `cmd` subcommand is now again
   optional and pypsh also supports to execute commands that are read from
   stdin.

   So the following invokations are possible::

    pypsh "myhosts[0-9]" "uptime"
    echo "uptime" | pypsh "myhosts[0-9]"
    pypsh "myhosts[0-9]" cmd "uptime"
    pypsh "myhosts[0-9]" copy /tmp/source /tmp/destination

   The `wait` and `serial` parameters have been merged into `interval`
   If the interval is zero the commands will be executed in parallel, if it's
   greater than zero the command/copy process will be serial with a sleep of
   `interval` after each host.

   In addition, the `show` subcommand has been removed. Simply use `uptime` or
   some other harmless command to show which hosts matched.

2013-06-24: 0.6.0
=================

 - **feature**: added new `--wait` parameter to the `cmd` subcommand.

2013-06-24: 0.5.0
=================

 - **feature**: added new `show` subcommand.

2013-05-16: 0.4.0
=================

 - **feature**: added new `copy` subcommand. Therefore the `cmd` command has
   to be specified explicitly and is no longer the default command.

2013-03-28: 0.3.0
=================

 - **feature**: added --serial (-s) parameter to run commands serial instead of
   parallel.

2013-03-01: 0.2.0
=================

 - **feature**: output is now colored

 - **feature**: a summary of matched hosts is printed before cmd execution

 - **feature**: a summary of (un)successful cmd invokations is printed after cmd
   execution

 - **bugfix**: improved error handling in case of wrong regex or ssh connection
   errors.

 - **bugfix**: fixed erroneous import in setup.py that caused it to require all
   dependencies to be already installed.

2013-02-12: 0.1.0
=================

 - initial version


pypsh.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import os
import sys
import re
import multiprocessing
from time import sleep
from paramiko import util, SSHConfig, SSHClient, WarningPolicy
from argh import ArghParser, command


class SSHExecutor(multiprocessing.Process):
    def __init__(self, host, cmd, config):
        super(SSHExecutor, self).__init__()
        self.host = host
        self.cmd = cmd
        self.config = config

    def stop(self):
        self.terminate()

    def run(self):
        exec_command(self.cmd, self.config, self.host)


@command
def pypsh(hostregex, cmd):
    config = SSHConfig()
    config.parse(open(os.path.expanduser('~/.ssh/config')))
    keys = util.load_host_keys(os.path.expanduser('~/.ssh/known_hosts'))
    hosts = []
    for key in keys:
        if re.match(hostregex, key):
            hosts.append(key)

    processes = []
    for host in hosts:
        p = SSHExecutor(host, cmd, config.lookup(host))
        p.start()
        processes.append(p)

    while multiprocessing.active_children():
        try:
            sleep(0.3)
        except KeyboardInterrupt:
            for p in processes:
                p.stop()
            sys.exit()


def exec_command(cmd, hostconfig, host):
    client = SSHClient()
    client.load_system_host_keys()
    client.set_missing_host_key_policy(WarningPolicy)
    client.connect(hostconfig.get('hostname'),
                   int(hostconfig.get('port', 22)),
                   username=hostconfig.get('user'))
    stdin, stdout, stderr = client.exec_command(cmd)
    for i, line in enumerate(stdout):
        line = line.rstrip()
        print("{0}: {1}".format(host, line))
    for i, line in enumerate(stderr):
        line = line.rstrip()
        print("{0}: {1}".format(host, line))
    client.close()


def main():
    p = ArghParser()
    p.set_default_command(pypsh)
    p.dispatch()


if __name__ == '__main__':
    main()


pypsh/main.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-

import sys
import os
import re
import multiprocessing
import select
import socket
from collections import defaultdict
from threading import Thread
from argparse import ArgumentParser
from time import sleep
from functools import partial
from paramiko import (util, SSHConfig, SSHClient, WarningPolicy,
                      BadHostKeyException, SSHException,
                      AuthenticationException as AuthException)
from termcolor import colored


class Printer(object):

    def __init__(self, host, stdout, stderr):
        def printer(stream, color):
            hostpart = colored("[{0}]".format(host), color) + ": "
            for line in stream:
                line = line.rstrip()
                print(hostpart + line)
        self.printers = [
            Thread(target=printer, args=(stdout, 'green')),
            Thread(target=printer, args=(stderr, 'red'))
        ]
        self.stdout = stdout
        self.stderr = stderr

    def loop(self):
        for p in self.printers:
            p.daemon = True
            p.start()
        for p in self.printers:
            p.join()
        if hasattr(self.stdout, 'close'):
            self.stdout.close()
        if hasattr(self.stderr, 'close'):
            self.stderr.close()


class Executor(multiprocessing.Process):
    """ abstract class that connects via ssh to a given host and executes
    `exec_command()`

    exec_command has to be overwritten in a subclass to do something useful.
    """

    def __init__(self, host, config):
        super(Executor, self).__init__()
        self.host = host
        self.config = config

    def stop(self):
        self.terminate()

    def run(self):
        sys.exit(self._exec())

    def exec_command(self):
        """overwrite this method to implement specific functionality

        this method has to return a tuple with 3 iterables.
        (stdout, stderr, channel)
        """

    def _exec(self):
        exitcode = 0
        client = SSHClient()
        client.load_system_host_keys()
        client.set_missing_host_key_policy(WarningPolicy())
        try:
            client.connect(self.config.get('hostname'),
                           int(self.config.get('port', 22)),
                           username=self.config.get('user'))
            stdout, stderr, channel = self.exec_command(client)
            Printer(self.host, stdout, stderr).loop()
            if channel:
                exitcode = channel.recv_exit_status()
        except IOError as e:
            print(colored('{0}: {1}'.format(self.host, str(e)), 'red'))
            exitcode = 1
        except (BadHostKeyException, AuthException, SSHException) as e:
            print(colored('{0}: {1}'.format(self.host, e.message)), 'red')
            exitcode = 1
        except Exception as e:
            print(colored('{0}: {1}'.format(self.host, e.message)), 'red')
            exitcode = 1
        finally:
            client.close()
            return exitcode


class SSHExecutor(Executor):
    """ execute a simple command via ssh """

    def __init__(self, host, config, cmd, pty):
        super(SSHExecutor, self).__init__(host, config)
        self.cmd = cmd
        self.get_pty = pty
        self.timeout = None

    def exec_command(self, client):
        chan = client.get_transport().open_session()
        if self.get_pty:
            chan.get_pty()
        chan.settimeout(self.timeout)
        chan.exec_command(self.cmd)
        stdin = chan.makefile('wb', -1)
        stdout = chan.makefile('r', -1)
        stderr = chan.makefile_stderr('r', -1)
        stdin.close()
        return stdout, stderr, chan


class CopyExecutor(Executor):
    """ copy a file from source to destination via ssh/sftp """

    def __init__(self, host, config, source, destination):
        super(CopyExecutor, self).__init__(host, config)
        self.source = source
        self.destination = destination

    def exec_command(self, client):
        sftp = client.open_sftp()
        sftp.put(self.source, self.destination)
        sftp.close()
        return (['Copied to {}'.format(self.host)], [], None)


RE_HOST_PORT = re.compile('\[(.*)\]:\d+.*')


def keys_from_config():
    with open(os.path.expanduser('~/.ssh/config'), 'r') as f:
        for line in f:
            if line.startswith('Host '):
                yield line[5:].strip()


def get_hosts(hostregex):
    """return all hosts that are in the known_hosts file and match the given
    regex.

    Will return an empty list if no hosts matched.
    """
    try:
        keys = util.load_host_keys(os.path.expanduser('~/.ssh/known_hosts'))
    except SSHException as e:
        print(colored('Error in ~/.ssh/known_hosts:', 'red', attrs=['bold']))
        sys.exit(colored(e.message, 'red'))
    hosts = []
    try:
        rex = re.compile(hostregex)
    except re.error:
        sys.exit(colored('Invalid regular expression!', 'red', attrs=['bold']))
    keys = [key for key in keys] + list(keys_from_config())
    for key in keys:
        custom_port_match = RE_HOST_PORT.match(key)
        if custom_port_match:
            key = custom_port_match.group(1)
        if rex.match(key):
            hosts.append(key)

    match_success = 'green' if len(hosts) > 0 else 'red'
    print('>>> {} Hosts matched:'.format(colored(len(hosts), match_success,
                                                 attrs=['bold'])))
    print('')
    print('\n'.join(sorted(hosts)))
    print('')
    return hosts


def print_result(processes):
    num_ok = sum([1 for p in processes if p.exitcode == 0])
    failures = sorted([p for p in processes if p.exitcode != 0],
                      key=lambda x: x.host)
    print('>>> {} successful invocations.'.format(colored(num_ok, 'green')))
    print('>>> {} failures:'.format(colored(len(failures), 'red')))
    for p in failures:
        print('\t{}'.format(p.host))


def start_procs(interval, hosts, starter_func):
    config = SSHConfig()
    config.parse(open(os.path.expanduser('~/.ssh/config')))

    processes = []
    for host in hosts:
        process = starter_func(host, config.lookup(host))
        process.start()
        if interval > 0.0:
            process.join()
            sleep(interval)
        processes.append(process)

    while multiprocessing.active_children():
        try:
            sleep(0.3)
        except KeyboardInterrupt:
            for p in processes:
                p.stop()
            break
    return processes


def cmd(hosts, cmd, interval=0.0, pty=False):
    print('>>> Starting to execute the command(s):')
    print('')
    cmd_executer = partial(SSHExecutor, cmd=cmd, pty=pty)
    processes = start_procs(interval, hosts, cmd_executer)
    print_result(processes)


def copy(source, hosts, destination, interval=0.0):
    if not os.path.isfile(source):
        print(colored('>>> Source {} does not exist'.format(source), 'red'))
        sys.exit(1)
    print('>>> Starting to copy the file:')
    print('')

    cmd_executer = partial(CopyExecutor,
                           source=source,
                           destination=destination)
    processes = start_procs(interval, hosts, cmd_executer)
    print_result(processes)


def interactive(args):
    hosts = get_hosts(args.hostregex)
    config = SSHConfig()
    config.parse(open(os.path.expanduser('~/.ssh/config')))

    clients = []
    transports = []
    channels = {}
    for host in hosts:
        client = SSHClient()
        client.load_system_host_keys()
        client.set_missing_host_key_policy(WarningPolicy())
        hostconfig = config.lookup(host)
        client.connect(hostconfig.get('hostname'),
                       int(hostconfig.get('port', 22)),
                       username=hostconfig.get('user'))
        clients.append(client)
        transport = client.get_transport()
        transports.append(transport)
        chan = transport.open_session()
        if args.pty:
            chan.get_pty()
        chan.settimeout(0.0)
        chan.invoke_shell()
        channels[chan] = host

    channels_stdin = list(channels.keys()) + [sys.stdin]
    print('Connections established.. waiting for input: ')
    lines = defaultdict(list)
    while True:
        try:
            sys.stdout.write(colored('> ', 'red', attrs=['bold']))
            sys.stdout.flush()
            r, w, e = select.select(channels_stdin, [], [])
            if sys.stdin in r:
                line = sys.stdin.readline()
                if line:
                    for i, chan in enumerate(channels):
                        try:
                            chan.send(line)
                        except socket.timeout:
                            print("couldn't send to channel {0}".format(i))
            else:
                for chan in channels:
                    if chan in r:
                        try:
                            while True:
                                x = chan.recv(1024).decode('utf-8')
                                lines[channels[chan]].append(x)
                                if len(x) == 0:
                                    break
                        except socket.timeout:
                            pass
                for host in lines:
                    print(colored(host, 'green'))
                    for c in lines[host]:
                        sys.stdout.write(c)
                    sys.stdout.flush()
                lines.clear()
        except KeyboardInterrupt:
            for chan in channels:
                chan.close()
            for t in transports:
                t.close()
            print('Bye')
            break


def dispatch(args):
    hosts = get_hosts(args.hostregex)
    if 'command' in args:
        cmd(hosts, args.command, args.interval, args.pty)
    else:
        copy(args.source, hosts, args.destination, args.interval)


def create_parser():
    parser = ArgumentParser(description='parallel ssh command execution')
    parser.add_argument('hostregex', type=str,
                        help='regular expression to match the hostnames')
    parser.add_argument(
        '-i', '--interval',
        type=float,
        default=0.0,
        help=('time to wait between command execution on the different hosts.'
              '0 means no wait time and everything is executed in parallel'))
    subparsers = parser.add_subparsers(help='sub-command help')

    cmd_subcommand = subparsers.add_parser(
        'cmd', help='execute a command on multiple hosts'
    )
    cmd_subcommand.set_defaults(func=dispatch)
    cmd_subcommand.add_argument('command', type=str, help='command to execute')
    cmd_subcommand.add_argument('--pty', action='count',
                                help='attach pty', default=False)

    copy_subcommand = subparsers.add_parser(
        'copy', help='copy a file from the local machine to multiple hosts'
    )
    copy_subcommand.set_defaults(func=dispatch)
    copy_subcommand.add_argument('source', type=str,
                                 help='path to the file on the local machine')
    copy_subcommand.add_argument('destination', type=str,
                                 help='path to the file on the remote hosts')

    interactive_cmd = subparsers.add_parser(
        'interactive', help='create an interactive connection to multiple hosts'
    )
    interactive_cmd.add_argument(
        '--pty', action='count', help='attach pty', default=False)
    interactive_cmd.set_defaults(func=interactive)
    return parser


def main():
    if len(sys.argv) == 2:
        hosts = get_hosts(sys.argv[1])
        if hosts:
            while sys.stdin in select.select([sys.stdin], [], [], 0)[0]:
                line = sys.stdin.readline()
                if line:
                    cmd(hosts, line, False)
                else:
                    sys.exit(0)
    elif len(sys.argv) == 3 and sys.argv[2] != 'interactive':
        # "pypsh <hostregex> <command>" should work too
        hosts = get_hosts(sys.argv[1])
        if hosts and sys.argv[2] != 'help':
            cmd(hosts, sys.argv[2], False)
            sys.exit(0)
    parser = create_parser()
    args = parser.parse_args()
    args.func(args)


if __name__ == '__main__':
    main()
